Add object detection model, dataset and training

seldauyanik-maxim opened the pull request:
Sample object detection dataset is added for street house numbers
detection problem.
Multi-box Loss is added for object detection training use.
Some utility functions for object detection models are added.
Some parameter specification for object deteciton using a yaml file
is added.
Training script is modified for object detection data loaders.
A Jupyter notebook is also added for visualization of the trained
model.

rotx-maxim commented on models/ai85net_adi_tinierssd.py:
```diff
@@ -0,0 +1,461 @@
+import ai8x
+import utils.object_detection_utils as obj_detect_utils
+
+sys.path.append('../')
```
Please ensure this is OS independent

seldauyanik-maxim replied:
Done and commit will be amended

rotx-maxim commented on models/ai85net_adi_tinierssd.py:
```diff
@@ -0,0 +1,461 @@
+
+sys.path.append('../')
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
This will conflict with the logic in train.py which is more complicated (e.g., supports --cpu)d

seldauyanik-maxim replied:
Done and commit will be amended

rotx-maxim commented on models/ai85net_adi_tinierssd.py:
```diff
@@ -0,0 +1,461 @@
+                                additional_scale = 1.
+                            prior_boxes.append([cx, cy, additional_scale, additional_scale])
+
+        prior_boxes = torch.FloatTensor(prior_boxes).to(device)  # (1700, 4)
```
You may have to pass in `device` or a tensor on that device to get the device from, or do the conversion on the return value of this function instead.

seldauyanik-maxim replied:
Done and commit will be amended

rotx-maxim commented on parsecmd.py:
```diff
@@ -235,6 +235,18 @@ def get_parser(model_names, dataset_names):
+                                    action='store_true',
+                                    help='Enable when training an Object Detection Model')
+
+    obj_detection_args.add_argument('--obj-detection-params', dest='obj_detection_params',
```
The `dest` is not needed here (it will convert - to _ automatically)

seldauyanik-maxim replied:
Done and commit will be amended

rotx-maxim commented on scripts/train_svhn_tinierssd.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+cd ..
```
Please remove `cd ..`. The scripts are expected to run as `scripts/xxxx.sh`

seldauyanik-maxim replied:
Done and commit will be amended

rotx-maxim commented on utils/object_detection_utils.py:
```diff
@@ -0,0 +1,296 @@
+# GitHub repo for the below helper methods:
+# https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Object-Detection:
+# MIT License
+# Copyright (c) 2019 Sagar Vinodababu
```
Please move this copyright to the top header, and add ours. See train.py for an example.

seldauyanik-maxim replied:
Done and commit will be amended

rotx-maxim reviewed (changes requested):
Please see comments above

rotx-maxim commented on losses/multiboxloss.py:
```diff
@@ -0,0 +1,183 @@
+        SSD paper (Liu, W. et al. (2016). SSD: Single Shot MultiBox Detector,
+        https://doi.org/10.1007/978-3-319-46448-0_2)
+
+        Multi Nox Loss code is from GitHub repo:
```
Please move the copyright into the file header as well just like you did for the other file

seldauyanik-maxim replied:
Sure done and will be commit amended

seldauyanik-maxim commented on models/ai85net_adi_tinierssd.py:
```diff
@@ -0,0 +1,468 @@
+import ai8x
+import utils.object_detection_utils as obj_detect_utils
+
+sys.path.append(os.path.join(os.getcwd(), '..'))
```
Correct, removed this one and will amend commit

seldauyanik-maxim commented on models/ai85net_adi_tinierssd.py:
```diff
@@ -0,0 +1,468 @@
+    Constructs a Tinier SSD model
+    """
+    assert not pretrained
+    # Can specify and provide different aspect ratios if required here
```
We can, but I do not think we should, Here is why:

For a changed aspect ratio, it will be more elegant to define another method e.g. tinierssd_for_egg(), call TinierSSD constructor inside with specific aspect ratios and add another entry to the models dictionary list to make it parametrized in training script call.

If we parametrize tinierssd() instead, then inside the training script where it just constructs model with the common kwargs, it should also specify aspect ratios  and may not be the best place.

seldauyanik-maxim commented on scripts/train_svhn_tinierssd.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+python train.py --deterministic --print-freq 200 --pr-curves --epochs 300 --optimizer Adam --lr 0.001 --wd 0 --model tinierssd --use-bias --momentum 0.9 --weight-decay 5e-4 --dataset SVHN_74 --device MAX78000 --obj-detection --obj-detection-params parameters/obj_detection_params_svhn.yaml --batch-size 16 --qat-policy policies/qat_policy_svhn.yaml --validation-split 0 --data /data/ "$@"
```
Sure, removed and will be amended

seldauyanik-maxim commented on utils/parse_obj_detection_yaml.py:
```diff
@@ -0,0 +1,50 @@
+    if parameters and 'multi_box_loss' not in parameters:
+        assert False, '`multi_box_loss` parameters must be defined'
+
+    if parameters['multi_box_loss'] and 'alpha' not in parameters['multi_box_loss']:
```
I see, the reason was the nested structure of the yaml entries and I wanted to make sure they are all set in given file. I refactored into a more readable version.

seldauyanik-maxim commented:
Dear all, I will an additional small commit with a couple of additions for training in multi-GPU systems, and will re-request your reviews:
- add .module for model attribute accesses
- updated Jupyter notebook for loading saved checkpoint
I also updated requirements (related to SVHN dataset's h5py requirement for reading .mat file)

rotx-maxim commented on losses/multiboxloss.py:
```diff
@@ -0,0 +1,182 @@
+        bounding boxes. SSD paper uses smooth L1 loss for this purpose and so as the below
+        implementation.
+
+        Confidence Loss measures the perfroamnce of the  object classification per box.Categorical
```
Spelling: performance

seldauyanik-maxim replied:
Fixed in next commit

rotx-maxim commented on datasets/svhn.py:
```diff
@@ -0,0 +1,489 @@
+                                         'digitStruct.mat')
+
+            if not os.path.exists(mat_file_path):
+                print('\nDownload the archieve file from: '
```
Add "Review the terms and conditions on http://ufldl.stanford.edu/housenumbers/ and then download..."

Spelling: **archive**

rotx-maxim commented on datasets/svhn.py:
```diff
@@ -0,0 +1,489 @@
+            if not os.path.exists(mat_file_path):
+                print('\nDownload the archieve file from: '
+                      'http://ufldl.stanford.edu/housenumbers/[train or test].tar.gz')
+                print('Extract the downloaded archive to path [data_dir]/SVHN.')
```
Change [data_dir] to actual data directory and elaborate what the final structure is supposed to look like.

MaximGorkem reviewed (approved):
Looks great. My only comment is that we need to optimize NMS paramaters and other training parameters like qat_start to increase the accuracy. These values are the first tried ones and need to be optimized.
